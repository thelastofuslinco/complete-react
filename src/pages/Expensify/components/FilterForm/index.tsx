import { ConnectedProps, connect } from 'react-redux'
import { FormHTMLAttributes } from 'react'
import { RootState, filterExpenses } from '../../../../store'
import Input from '../../../../components/Input'
import Select from '../../../../components/Select'
import DatePicker from '../../../../components/DatePicker'

interface Props extends PropsFromRedux, FormHTMLAttributes<HTMLFormElement> {}

const FilterForm = ({ filters, filterExpenses, ...props }: Props) => {
  const handleSubmit = (event) => {
    event.preventDefault()
    const { text, sort, date } = event.target

    const startDate = date.value
      ? new Date(date.value.split(' - ')[0]).toISOString()
      : null
    const endDate = date.value
      ? new Date(date.value.split(' - ')[1]).toISOString()
      : null

    filterExpenses({
      startDate,
      endDate,
      text: text.value,
      sortBy: sort.value
    })
  }

  return (
    <form onSubmit={handleSubmit} {...props}>
      <Input name="text" value={filters.text} />
      <Select name="sort" value={filters.sortBy}>
        <option value="amount asc">amount asc</option>
        <option value="amount desc">amount desc</option>
        <option value="description asc">description asc</option>
        <option value="description desc">description desc</option>
        <option value="createdAt asc">createdAt asc</option>
        <option value="createdAt desc">createdAt desc</option>
      </Select>
      <DatePicker maxDate={new Date()} />
      <button type="submit">send</button>
    </form>
  )
}

const conector = connect((state: RootState) => state.expenses, {
  filterExpenses: filterExpenses
})

type PropsFromRedux = ConnectedProps<typeof conector>

export default conector(FilterForm)
