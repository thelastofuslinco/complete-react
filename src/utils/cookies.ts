export function setCookie(name, value, days) {
  const d = new Date()
  d.setTime(d.getTime() + days * 24 * 60 * 60 * 1000)
  const expires = 'expires=' + d.toUTCString()
  document.cookie = `${name}=${value};${expires};path=/;Secure;HttpOnly;SameSite=Strict`
}

export function getCookie(cookieName) {
  const name = cookieName + '='
  const ca = document.cookie.split(';')

  for (let i = 0; i < ca.length; i++) {
    let c = ca[i]
    while (c.charAt(0) == ' ') {
      c = c.substring(1)
    }
    if (c.indexOf(name) == 0) {
      return c.substring(name.length, c.length)
    }
  }
  return ''
}
