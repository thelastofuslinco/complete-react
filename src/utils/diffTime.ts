export const diffTime = (time) => {
  const date1 = new Date(time).getTime()
  const date2 = new Date().getTime()
  const diff = Math.abs(date2 - date1)
  const diffTime = {
    s: Math.floor(diff / 1000),
    m: Math.floor(diff / (1000 * 60)),
    h: Math.floor(diff / (1000 * 60 * 60))
  }

  if (diffTime.h > 1) {
    return `${diffTime.h} hour ago`
  } else if (diffTime.m > 1) {
    return `${diffTime.m} minutes ago`
  } else {
    return `${diffTime.s} seconds ago`
  }
}
