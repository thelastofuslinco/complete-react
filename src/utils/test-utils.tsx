import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import expensesSlice from '../store/slices/expensesSlice'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { expenses: expensesSlice },
      preloadedState
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions })
  }
}
