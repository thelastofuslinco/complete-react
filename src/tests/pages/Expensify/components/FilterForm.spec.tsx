import React from 'react'
import { screen } from '@testing-library/react'
import FilterForm from '../../../../pages/Expensify/components/FilterForm'
import { renderWithProviders } from '../../../../utils/test-utils'

describe('Filter form component', () => {
  test('should render filter form', () => {
    renderWithProviders(<FilterForm name="" />)
    const element = screen.getByRole('form', { name: '' })
    expect(element).toMatchSnapshot()
  })
})
