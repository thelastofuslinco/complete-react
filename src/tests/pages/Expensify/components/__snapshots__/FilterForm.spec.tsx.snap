// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter form component should render filter form 1`] = `
<form
  data=""
  name=""
>
  <input
    class="input undefined"
    name="text"
    value=""
  />
  <select
    class="select undefined"
    name="sort"
  >
    <option
      value="amount asc"
    >
      amount asc
    </option>
    <option
      value="amount desc"
    >
      amount desc
    </option>
    <option
      value="description asc"
    >
      description asc
    </option>
    <option
      value="description desc"
    >
      description desc
    </option>
    <option
      value="createdAt asc"
    >
      createdAt asc
    </option>
    <option
      value="createdAt desc"
    >
      createdAt desc
    </option>
  </select>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        class=""
        name="date"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    type="submit"
  >
    send
  </button>
</form>
`;
